docs/source/upgrade.rst

```diff
@@ -1,819 +1,124 @@
+Argument validation of functions is now strictened to check device consistency of argument variables to provide better error messages to users.
+Suppose the following code:
 
-In Chainer v2, the concept of *training mode* is added.
-It is represented by a thread-local flag ``chainer.config.train``, which is a part of :ref:`the unified configuration <configuration>`.
-When ``chainer.config.train`` is ``True``, functions of Chainer run in the training mode, and otherwise they run in the test mode.
-For example, :class:`~links.BatchNormalization` and :func:`~functions.dropout` behave differently in each mode.
+.. code-block:: py
 
-In Chainer v1, such a behavior was configured by the ``train`` or ``test`` argument of each function.
-**This train/test argument has been removed in Chainer v2.**
-If your code is using the ``train`` or ``test`` argument, you have to update it.
-In most cases, what you have to do is just removing the ``train`` / ``test`` argument from any function calls.
+   v1 = chainer.Variable(np.arange(10, dtype=np.float32))      # CPU
+   v2 = chainer.Variable(cupy.arange(10, dtype=cupy.float32))  # GPU
 
-.. admonition:: Example
+   # The line below raises an exception, because arguments are on different device.
+   F.maximum(v1, v2)
 
-   Consider the following model definition and the code to call it in test mode written for Chainer v1.
+Prior to v4, the above code raises an exception like ``ValueError: object __array__ method not producing an array``, which was difficult to understand.
+In v4, the error message would become ``ValueError: numpy and cupy arrays are mixed in the forward input (Maximum)``.
+This kind of error usually occur by mistake (for example, not performing ``to_gpu`` for some variables).
 
-   .. code-block:: py
+.. attention::
 
-      # Chainer v1
-      import chainer.functions as F
+   As the argument validation is strictened, call of new-style functions (inheriting :class:`FunctionNode` class) intentionally mixing NumPy/CuPy arrays in arguments will not work in Chainer v4.
+   Please transfer all arrays to the same device before calling functions.
 
-      class MyModel(chainer.Link):
-          ...
+Note that old-style function classes (inheriting :class:`Function`) are not affected.
 
-          def __call__(self, x, train=True):
-              return f(F.dropout(x, train=train))
+Reproducibility Improvement with MultiProcessIterator
+-----------------------------------------------------
 
-      m = MyModel(...)
-      y = m(x, train=False)
+Prior to v4, NumPy random states are shared between threads when using :class:`chainer.iterators.MultiprocessIterator`.
+This release fixed this issue by using a separate random state for a sub thread to improve reproducibility of experiments.
+See the discussion in `#3575 <https://github.com/chainer/chainer/pull/3575>`_ for more details.
 
-   In Chainer v2, it should be updated into the following code:
+Update of Docker Images
```

Comment:
ditto. (Do we need this topic in upgrade guide?)

Reply:
I think this is needed, because CUDA 7.5 users can no longer run the official docker image.

Reply:
I mean, users with older NVIDIA driver version cannot run the Docker image, although we still continue to support CUDA 7.5.
https://github.com/NVIDIA/nvidia-docker/wiki/CUDA#requirements